lib/backend/pgbk/pgbk.go

```diff
@@ -0,0 +1,580 @@
+	}
+	if c.ExpiryBatchSize < 0 {
+		return trace.BadParameter("expiry batch size must be non-negative")
+	}
+	if c.ExpiryBatchSize == 0 {
+		c.ExpiryBatchSize = defaultExpiryBatchSize
+	}
+
+	return nil
+}
+
+// NewFromParams starts and returns a [*Backend] with the given params
+// (generally read from the Teleport configuration file).
+func NewFromParams(ctx context.Context, params backend.Params) (*Backend, error) {
+	var cfg Config
+	if err := utils.ObjectToStruct(params, &cfg); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	bk, err := NewWithConfig(ctx, cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return bk, nil
+}
+
+// NewWithConfig starts and returns a [*Backend] with the given [Config].
+func NewWithConfig(ctx context.Context, cfg Config) (*Backend, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	poolConfig, err := pgxpool.ParseConfig(cfg.ConnString)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	log := logrus.WithField(trace.Component, componentName)
+
+	if cfg.AuthMode == AzureADAuth {
+		bc, err := pgcommon.AzureBeforeConnect(cfg.AzureClientID, log)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		poolConfig.BeforeConnect = bc
+	}
+
+	poolConfig.AfterConnect = func(ctx context.Context, c *pgx.Conn) error {
+		_, err := c.Exec(ctx, "SET default_transaction_isolation TO serializable", pgx.QueryExecModeExec)
+		return trace.Wrap(err)
+	}
+
+	log.Info("Setting up backend.")
+
+	pgcommon.TryEnsureDatabase(ctx, poolConfig, log)
+
+	pool, err := pgxpool.NewWithConfig(ctx, poolConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	b := &Backend{
+		cfg:    cfg,
+		log:    log,
+		pool:   pool,
+		buf:    backend.NewCircularBuffer(),
+		cancel: cancel,
+	}
+
+	if err := b.setupAndMigrate(ctx); err != nil {
+		b.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	if !cfg.DisableExpiry {
+		b.wg.Add(1)
+		go func() {
+			defer b.wg.Done()
+			b.backgroundExpiry(ctx)
+		}()
+	}
+
+	b.wg.Add(1)
+	go func() {
+		defer b.wg.Done()
+		b.backgroundChangeFeed(ctx)
+	}()
+
+	return b, nil
+}
+
+// Backend is a PostgreSQL-backed [backend.Backend].
+type Backend struct {
+	cfg  Config
+	log  logrus.FieldLogger
+	pool *pgxpool.Pool
+	buf  *backend.CircularBuffer
+
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+func (b *Backend) Close() error {
+	b.cancel()
+	b.wg.Wait()
+	b.buf.Close()
+	b.pool.Close()
+	return nil
+}
+
+var schemas = []string{
+	`CREATE TABLE kv (
+		key bytea NOT NULL,
+		value bytea NOT NULL,
+		expires timestamptz,
+		revision uuid NOT NULL,
+		CONSTRAINT kv_pkey PRIMARY KEY (key)
+	);
+	CREATE INDEX kv_expires_idx ON kv (expires) WHERE expires IS NOT NULL;`,
+}
+
+// setupAndMigrate sets up the database schema, applying the migrations in the
+// [schemas] slice starting from the first non-applied one.
+func (b *Backend) setupAndMigrate(ctx context.Context) error {
```

Comment:
I see this code is very similar to https://github.com/gravitational/teleport/pull/29057/files#diff-4a9bdc66c1efc75d6078bc6616215a6b6784458b655f395bebcac293e69c852fR267

1. Do we want to keep versioning per table? I assume we need that as backend and events don't need to use Postgres?
2. Shouldn't we make that logic more generic/common?

Reply:
Made the logic generic in 04043cc1cffa246aa6e4404488bfe02d92030a65.

We should use different table names for pgbk and pgevents because if they're pointed to the same database one of the two will assume that the schema is already correct even though tables are missing, so we'll get very strange errors after a supposedly successful initialization.